### PR TITLE
MGDOBR-1259: Processor: API [C]RUD operation

### PR DESCRIPTION
**Pull request by manstis**
[https://issues.redhat.com/browse/MGDOBR-1259](https://issues.redhat.com/browse/MGDOBR-1259)

## PR owner checklist

Please make sure that your PR meets the following requirements:

- [x] Your code is properly formatted according to [this configuration](https://github.com/kiegroup/kogito-runtimes/tree/main/kogito-build/kogito-ide-config).  
  *Run `mvn clean verify -DskipTests` so that imports are correctly set.*
- [x] Your commit messages are clear and reference the JIRA issue e.g: "[MGDOBR-XXX] - $clear_explanation_of_what_you_did"
- [x] The layers in the `kustomize` folder have been updated accordingly.
- [ ] All new functionality is tested
- [x] Pull Request title is properly formatted: `MGDOBR-XYZ Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains link to any dependent or related Pull Request
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket

## PR reviewer(s) checklist

- [ ] Check if code and Github action workflows have been modified and if the modification is safe
- [ ] If the modification is safe, add the `safe to test` label

<details>
<summary>
How to trigger pipelines and use the bots:
</summary>

- <b>Run the end to end pipeline</b>  
  Annotate the pull request with the label: `safe to test`. If you want to run the pipeline again, remove and add it again.

- <b>Rebase the pull request</b>  
  Comment with: `/rebase`.
- <b>Deploy BOT</b>
  Comment with `/deploy <dev,stable> when the PR has been merged to deploy to a target environment.

</details>

### CODE REVIEW

**Comment by manstis** [comment]
@lucamolteni This is FYI as it includes some parts of interest to the Data Plane.

File: infra-parent/infra-v2/src/main/java/com/redhat/service/smartevents/infra/v2/api/models/ConditionStatus.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
+package com.redhat.service.smartevents.infra.v2.api.models;
+
+import com.fasterxml.jackson.annotation.JsonValue;
+import com.redhat.service.smartevents.infra.core.models.BaseEnumeration;
+
+public enum ConditionStatus implements BaseEnumeration {
```
Comment (manstis) [inline infra-parent/infra-v2/src/main/java/com/redhat/service/smartevents/infra/v2/api/models/ConditionStatus.java:6]: There have been issues with Jackson serialisation vs what we advertise in OpenAPI.

This approach fixes it (and is what we do on other enums).

File: infra-parent/infra-v2/src/main/java/com/redhat/service/smartevents/infra/v2/api/models/DefaultConditions.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
+package com.redhat.service.smartevents.infra.v2.api.models;
+
+import java.util.List;
+
+import com.redhat.service.smartevents.infra.v2.api.models.dto.ConditionDTO;
+
+public final class DefaultConditions {
+
+    public static final String CP_DATA_PLANE_READY_NAME = "DataPlaneReady";
```
Comment (manstis) [inline infra-parent/infra-v2/src/main/java/com/redhat/service/smartevents/infra/v2/api/models/DefaultConditions.java:9]: @lucamolteni "Common" (well defined) names for `Condition`'s available to both `manager` and `operator`.

File: infra-parent/infra-v2/src/main/java/com/redhat/service/smartevents/infra/v2/api/models/dto/ConditionDTO.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+import java.time.ZonedDateTime;
+import java.util.Objects;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.redhat.service.smartevents.infra.v2.api.models.ConditionStatus;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class ConditionDTO {
```
Comment (manstis) [inline infra-parent/infra-v2/src/main/java/com/redhat/service/smartevents/infra/v2/api/models/dto/ConditionDTO.java:11]: @r00ta @lucamolteni "base" `Condition` class that can be used to/from `manager`/`operator`.

File: infra-parent/infra-v2/src/main/java/com/redhat/service/smartevents/infra/v2/api/models/processors/ProcessorDefinition.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+
+public class ProcessorDefinition {
+
+    @JsonProperty("flows")
+    @NotNull(message = "Processor flows cannot be null")
+    // ObjectNode is not rendered properly by swagger
+    @Schema(implementation = Object.class, required = true)
+    private ObjectNode flows;
```
Comment (manstis) [inline infra-parent/infra-v2/src/main/java/com/redhat/service/smartevents/infra/v2/api/models/processors/ProcessorDefinition.java:18]: @r00ta "wrapper" for Camel's `flows` (useful for if we need to persist additional meta-data later).

File: manager-parent/manager-v2/src/main/java/com/redhat/service/smartevents/manager/v2/metrics/MetricsServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+import com.redhat.service.smartevents.infra.v2.api.models.OperationType;
+import com.redhat.service.smartevents.manager.v2.persistence.models.ManagedResourceV2;
+import com.redhat.service.smartevents.manager.v2.persistence.models.Operation;
+import com.redhat.service.smartevents.manager.v2.services.UserUtilities;
+
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Tag;
+
+@ApplicationScoped
+public class MetricsServiceImpl implements ManagerMetricsServiceV2 {
```
Comment (manstis) [inline manager-parent/manager-v2/src/main/java/com/redhat/service/smartevents/manager/v2/metrics/MetricsServiceImpl.java:26]: Untried or tested.. copied from `v1` with some hackery for `v2`. Untried or tested yet.. but needed.

File: manager-parent/manager-v2/src/main/java/com/redhat/service/smartevents/manager/v2/metrics/MetricsServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+            case MANAGER_RESOURCE_MODIFY:
+                return Duration.between(assertUpdateOperation(managedResource, operation), ZonedDateTime.now(ZoneOffset.UTC));
+            case MANAGER_RESOURCE_DELETE:
+                return Duration.between(assertDeleteOperation(managedResource, operation), ZonedDateTime.now(ZoneOffset.UTC));
+            default:
+                throw new IllegalStateException(String.format("Unable to calculate operation duration for MetricsOperation '%s'", operation));
+        }
+    }
+
+    private ZonedDateTime assertUpdateOperation(ManagedResourceV2 managedResource, MetricsOperation metricsOperation) {
```
Comment (manstis) [inline manager-parent/manager-v2/src/main/java/com/redhat/service/smartevents/manager/v2/metrics/MetricsServiceImpl.java:110]: Hackery to calculate the duration of `UPDATE` and `DELETE` operations.

File: manager-parent/manager-v2/src/main/java/com/redhat/service/smartevents/manager/v2/services/UserUtilities.java [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+import com.redhat.service.smartevents.infra.core.models.ManagedResourceStatus;
+import com.redhat.service.smartevents.infra.v2.api.models.ComponentType;
+import com.redhat.service.smartevents.infra.v2.api.models.OperationType;
+import com.redhat.service.smartevents.manager.v2.persistence.models.Bridge;
+import com.redhat.service.smartevents.manager.v2.persistence.models.Condition;
+import com.redhat.service.smartevents.manager.v2.persistence.models.ManagedResourceV2;
+import com.redhat.service.smartevents.manager.v2.persistence.models.Operation;
+import com.redhat.service.smartevents.manager.v2.persistence.models.Processor;
+
+public class UserUtilities {
```
Comment (manstis) [inline manager-parent/manager-v2/src/main/java/com/redhat/service/smartevents/manager/v2/services/UserUtilities.java:17]: @r00ta We know this has been refactored and moved in your local.

File: manager-parent/manager-v2/src/main/java/com/redhat/service/smartevents/manager/v2/workers/Work.java [hunk 8]
```diff
@@ -0,0 +1,64 @@
+package com.redhat.service.smartevents.manager.v2.workers;
+
+import java.time.ZoneOffset;
+import java.time.ZonedDateTime;
+
+import com.redhat.service.smartevents.manager.core.models.ManagedResource;
+
+public class Work {
```
Comment (manstis) [inline manager-parent/manager-v2/src/main/java/com/redhat/service/smartevents/manager/v2/workers/Work.java:8]: Temporary to support mocked `Work` use in service class.

File: manager-parent/manager-v2/src/main/java/com/redhat/service/smartevents/manager/v2/workers/WorkManager.java [hunk 9]
```diff
@@ -0,0 +1,36 @@
+package com.redhat.service.smartevents.manager.v2.workers;
+
+import com.redhat.service.smartevents.manager.core.models.ManagedResource;
+
+public interface WorkManager {
```
Comment (manstis) [inline manager-parent/manager-v2/src/main/java/com/redhat/service/smartevents/manager/v2/workers/WorkManager.java:5]: Temporary to support mocked `Work` use in service class.

File: manager-parent/manager-v2/src/main/java/com/redhat/service/smartevents/manager/v2/workers/quartz/QuartzWorkConvertor.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
+package com.redhat.service.smartevents.manager.v2.workers.quartz;
+
+import java.time.ZonedDateTime;
+
+import org.quartz.JobDataMap;
+
+import com.redhat.service.smartevents.manager.v2.workers.Work;
+
+public class QuartzWorkConvertor {
```
Comment (manstis) [inline manager-parent/manager-v2/src/main/java/com/redhat/service/smartevents/manager/v2/workers/quartz/QuartzWorkConvertor.java:9]: Temporary to support mocked `Work` use in service class.

File: manager-parent/manager-v2/src/main/java/com/redhat/service/smartevents/manager/v2/workers/quartz/QuartzWorkManagerImpl.java [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+import com.redhat.service.smartevents.manager.core.models.ManagedResource;
+import com.redhat.service.smartevents.manager.v2.workers.Work;
+import com.redhat.service.smartevents.manager.v2.workers.WorkManager;
+
+import io.quarkus.runtime.Quarkus;
+
+import static com.redhat.service.smartevents.manager.v2.workers.quartz.QuartzWorkConvertor.convertToJobData;
+
+@ApplicationScoped
+public class QuartzWorkManagerImpl implements WorkManager {
```
Comment (manstis) [inline manager-parent/manager-v2/src/main/java/com/redhat/service/smartevents/manager/v2/workers/quartz/QuartzWorkManagerImpl.java:32]: Temporary to support mocked `Work` use in service class.

File: manager-parent/manager-v2/src/main/java/com/redhat/service/smartevents/manager/v2/workers/quartz/WorkJob.java [hunk 12]
```diff
@@ -0,0 +1,13 @@
+package com.redhat.service.smartevents.manager.v2.workers.quartz;
+
+import org.quartz.Job;
+import org.quartz.JobExecutionContext;
+
+public class WorkJob implements Job {
```
Comment (manstis) [inline manager-parent/manager-v2/src/main/java/com/redhat/service/smartevents/manager/v2/workers/quartz/WorkJob.java:6]: Temporary to support mocked `Work` use in service class.

File: manager-parent/manager-v2/src/main/java/com/redhat/service/smartevents/manager/v2/persistence/models/Condition.java [hunk 13]
```diff
@@ -12,80 +12,84 @@
 import javax.persistence.Table;
 
 import com.redhat.service.smartevents.infra.v2.api.models.ComponentType;
+import com.redhat.service.smartevents.infra.v2.api.models.ConditionStatus;
+import com.redhat.service.smartevents.infra.v2.api.models.dto.ConditionDTO;
 
 @Entity
 @Table(name = "CONDITION")
-public class Condition {
+public class Condition extends ConditionDTO {
```
[the remark this answers is not in the document]
Comment (lucamolteni) [reply manager-parent/manager-v2/src/main/java/com/redhat/service/smartevents/manager/v2/persistence/models/Condition.java:20]: This looks very weird. DTOs are meant to be copy of the classes, are we sure about this?

File: manager-parent/manager-v2/src/main/java/com/redhat/service/smartevents/manager/v2/services/UserUtilities.java [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+
+    private static ManagedResourceStatus getManagedResourceStatus(List<Condition> conditions, Operation operation) {
+        switch (operation.getType()) {
+            case CREATE:
+            case UPDATE:
+                if (Objects.isNull(conditions) || conditions.isEmpty()) {
+                    return ManagedResourceStatus.ACCEPTED;
+                }
+                // The ordering of these checks is important!
+                if (conditions.stream().anyMatch(c -> c.getStatus().equals(CONDITION_STATUS_FAILED))) {
```
Comment (lucamolteni) [inline manager-parent/manager-v2/src/main/java/com/redhat/service/smartevents/manager/v2/services/UserUtilities.java:57]: I think we should rewrite this part of the code for clarity

**Review by lucamolteni — CHANGES_REQUESTED** [review]

File: manager-parent/manager-v2/src/main/java/com/redhat/service/smartevents/manager/v2/api/ManagerConstraintViolationExceptionMapper.java [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+import javax.ws.rs.ext.Provider;
+
+import com.redhat.service.smartevents.infra.core.exceptions.BridgeErrorService;
+import com.redhat.service.smartevents.infra.core.exceptions.mappers.ConstraintViolationExceptionMapper;
+import com.redhat.service.smartevents.infra.core.models.responses.ErrorResponse;
+import com.redhat.service.smartevents.infra.v2.api.V2APIConstants;
+
+@Provider
+@ApplicationScoped
+public class ManagerConstraintViolationExceptionMapper extends ConstraintViolationExceptionMapper {
```
Comment (manstis) [inline manager-parent/manager-v2/src/main/java/com/redhat/service/smartevents/manager/v2/api/ManagerConstraintViolationExceptionMapper.java:16]: @r00ta Added... but probably affected by your refactoring.

It's presence here almost certainly guarantees I duplicate errors in the responses when running `v1` and `v2` together.

File: manager-parent/manager-v2/src/main/java/com/redhat/service/smartevents/manager/v2/persistence/models/Condition.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (manstis) [reply manager-parent/manager-v2/src/main/java/com/redhat/service/smartevents/manager/v2/persistence/models/Condition.java:20]: Well.... ordinarily a DTO represents a simplified domain object to support serialisation of data etc. The domain class comes first; then the need to serialise some of its data elsewhere. However in this case the relationship is the other way around.. the DTO has come first; then my domain object extends it to enrich it with more data that we need to store in the Control Plane. Arguably if all development was first based on data transfer all domain objects would extend the DTO... but that's best argued over a beer :beers: 

@r00ta originally had this class duplicate most of the DTO but with a couple of extra fields we need stored.

I'm OK to refactor back to duplicating code if that's the preference.

**Comment by manstis** [comment]
@r00ta @ChughShilpa @winklerm This is ready for review now.

It has "more" than it probably needs. @r00ta took the decision to comment blocks out for `Bridge`'s whereas I added the minimal implementations (of required services etc).. If you'd prefer for me to adopt @r00ta approach please say. Happy to do so.. I am an advocate for consistency!

**Comment by manstis** [comment]
@lucamolteni I kept you as a reviewer for your opinion of the `ConditionDTO`.... I know you objected in my draft.

File: manager-parent/manager-core/src/main/resources/db/migration/0.2/V0.2.1__create_v2_db.sql [hunk 16]
```diff
@@ -30,7 +30,7 @@ CREATE TABLE PROCESSOR_V2 (
     id                             varchar(255) NOT NULL PRIMARY KEY,
 
     name                           varchar(255) NOT NULL,
-    flows                          jsonb        NOT NULL,
+    definition                     jsonb        NOT NULL,
```
Comment (r00ta) [inline manager-parent/manager-core/src/main/resources/db/migration/0.2/V0.2.1__create_v2_db.sql:33]: you need a new migration in order to apply this change

File: manager-parent/manager-v2/src/main/java/com/redhat/service/smartevents/manager/v2/api/ManagerConstraintViolationExceptionMapper.java [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+        super.init();
+    }
+
+    @Override
+    protected ErrorResponse toErrorResponse(ConstraintViolation<?> cv) {
+        ErrorResponse errorResponse = super.toErrorResponse(cv);
+        errorResponse.setHref(V2APIConstants.V2_ERROR_API_BASE_PATH + errorResponse.getId());
+        return errorResponse;
+    }
+}
```
Comment (r00ta) [inline manager-parent/manager-v2/src/main/java/com/redhat/service/smartevents/manager/v2/api/ManagerConstraintViolationExceptionMapper.java:35]: Does it work outside the manager-v2 package (for example in the final `manager` application that combines v1 and v2)?

File: manager-parent/manager-v2/src/main/java/com/redhat/service/smartevents/manager/v2/api/ManagerInternalPlatformExceptionMapper.java [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+import javax.ws.rs.ext.Provider;
+
+import com.redhat.service.smartevents.infra.core.exceptions.BridgeErrorService;
+import com.redhat.service.smartevents.infra.core.exceptions.definitions.platform.InternalPlatformException;
+import com.redhat.service.smartevents.infra.core.exceptions.mappers.InternalPlatformExceptionMapper;
+import com.redhat.service.smartevents.infra.core.models.responses.ErrorResponse;
+import com.redhat.service.smartevents.infra.v2.api.V2APIConstants;
+
+@Provider
+@ApplicationScoped
```
Comment (r00ta) [inline manager-parent/manager-v2/src/main/java/com/redhat/service/smartevents/manager/v2/api/ManagerInternalPlatformExceptionMapper.java:15]: same here

File: manager-parent/manager-v2/src/main/java/com/redhat/service/smartevents/manager/v2/metrics/MetricsServiceImpl.java [hunk 19]
```diff
@@ -0,0 +1,136 @@
...
+
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Tag;
+
+import static com.redhat.service.smartevents.manager.v2.utils.StatusUtilities.getManagedResourceStatus;
+
+@ApplicationScoped
+public class MetricsServiceImpl implements ManagerMetricsServiceV2 {
+
+    /*
```
Comment (r00ta) [inline manager-parent/manager-v2/src/main/java/com/redhat/service/smartevents/manager/v2/metrics/MetricsServiceImpl.java:29]: is this going to generate duplicated metrics (for v1 and v2)? Or is it using different tags?

File: manager-parent/manager-v2/src/main/java/com/redhat/service/smartevents/manager/v2/services/ProcessorServiceImpl.java [hunk 20]
```diff
@@ -0,0 +1,144 @@
...
+    @Inject
+    ProcessorDAO processorDAO;
+
+    @Inject
+    BridgeService bridgeService;
+
+    @Inject
+    WorkManager workManager;
+
+    @Inject
```
Comment (r00ta) [inline manager-parent/manager-v2/src/main/java/com/redhat/service/smartevents/manager/v2/services/ProcessorServiceImpl.java:55]: Missing @V2 annotation

File: manager-parent/manager-v2/src/main/java/com/redhat/service/smartevents/manager/v2/persistence/models/Condition.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (manstis) [reply manager-parent/manager-v2/src/main/java/com/redhat/service/smartevents/manager/v2/persistence/models/Condition.java:20]: @lucamolteni The vote is 1-1 at the moment.. you have the power of the deciding decision!!!

File: manager-parent/manager-core/src/main/resources/db/migration/0.2/V0.2.1__create_v2_db.sql (hunk 16 above)
Comment (manstis) [reply manager-parent/manager-core/src/main/resources/db/migration/0.2/V0.2.1__create_v2_db.sql:33]: Done.

File: manager-parent/manager-v2/src/main/java/com/redhat/service/smartevents/manager/v2/api/ManagerConstraintViolationExceptionMapper.java (hunk 17 above)
Comment (manstis) [reply manager-parent/manager-v2/src/main/java/com/redhat/service/smartevents/manager/v2/api/ManagerConstraintViolationExceptionMapper.java:35]: That is my fear.. that we'd have an `ExceptionMapper` for both `v1` and a `v2`. Both would map the same exceptions and we'd end up duplicating every mapped error.. Perhaps I'll remove this and edit the test until Errors has been redesigned.

File: manager-parent/manager-v2/src/main/java/com/redhat/service/smartevents/manager/v2/api/ManagerInternalPlatformExceptionMapper.java (hunk 18 above)
Comment (manstis) [reply manager-parent/manager-v2/src/main/java/com/redhat/service/smartevents/manager/v2/api/ManagerInternalPlatformExceptionMapper.java:15]: Same here too.

Comment (manstis) [reply manager-parent/manager-v2/src/main/java/com/redhat/service/smartevents/manager/v2/api/ManagerInternalPlatformExceptionMapper.java:15]: Same here too.

File: manager-parent/manager-v2/src/main/java/com/redhat/service/smartevents/manager/v2/api/ManagerConstraintViolationExceptionMapper.java (hunk 17 above)
Comment (manstis) [reply manager-parent/manager-v2/src/main/java/com/redhat/service/smartevents/manager/v2/api/ManagerConstraintViolationExceptionMapper.java]: Removed until Error Handling refactor is complete.

File: manager-parent/manager-v2/src/main/java/com/redhat/service/smartevents/manager/v2/api/ManagerInternalPlatformExceptionMapper.java (hunk 18 above)
Comment (manstis) [reply manager-parent/manager-v2/src/main/java/com/redhat/service/smartevents/manager/v2/api/ManagerInternalPlatformExceptionMapper.java]: Removed until Error Handling refactor is complete.

File: manager-parent/manager-v2/src/main/java/com/redhat/service/smartevents/manager/v2/metrics/MetricsServiceImpl.java (hunk 19 above)
Comment (manstis) [reply manager-parent/manager-v2/src/main/java/com/redhat/service/smartevents/manager/v2/metrics/MetricsServiceImpl.java]: Removed until Error Handling refactor is complete.

File: manager-parent/manager-v2/src/main/java/com/redhat/service/smartevents/manager/v2/persistence/models/Condition.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (manstis) [reply manager-parent/manager-v2/src/main/java/com/redhat/service/smartevents/manager/v2/persistence/models/Condition.java]: The final score was 2-1. I have reverted this to a standalone POJO with no inheritance.

File: manager-parent/manager-core/src/main/resources/db/migration/0.2/V0.2.1__create_v2_db.sql [hunk 21]
```diff
@@ -1,66 +1,69 @@
-CREATE TABLE BRIDGE_V2 (
```
Comment (r00ta) [inline manager-parent/manager-core/src/main/resources/db/migration/0.2/V0.2.1__create_v2_db.sql:1]: revert the changes in this file before the merge otherwise you'll make flyway cry ;)

**Review by r00ta — APPROVED** [review]
Great job as usual :+1:

**Review by winklerm — APPROVED** [review]
